mars/core.py

```diff
@@ -432,6 +443,32 @@ def build_graph(self, graph=None, cls=DAG, tiled=False, compose=True, executed_k
                           if c not in visited])
         if tiled and compose:
             graph.compose(keys=keys)
+
+        def adapt_tensor(tensor):
+            new_tensor = tensor.copy()
+            new_tensor.chunks = None
+
+            if len(graph.predecessors(tensor)) == 0:
+                new_tensor.update_params({'raw_chunk_size': node.nsplits})
+            new_tensor.update_id(tensor.id)
+            return new_tensor
+
+        if not tiled and has_tiled_tensor:
```

Comment:
I think it is much cleaner to write ``any(n.is_coarse() for n in graph)`` instead of introducing another flag

Reply:
Just don't want to extra traversal.

Reply:
Looping over nodes in a graph is not traversing, it is only enumerating over a set and the overhead is quite light.

Reply:
Maybe just add a `to_coarse` method, and if `tiled` is False, just call `to_coarse` when adding a node to graph could be better.